api/testhelpers/proxy.go

```diff
@@ -98,3 +101,55 @@ func (p *ProxyHandler) Count() int {
 	defer p.Unlock()
 	return p.count
 }
+
+// Reset sets the counter for proxied requests to zero.
+func (p *ProxyHandler) Reset() {
+	p.Lock()
+	defer p.Unlock()
+	p.count = 0
+}
+
+// GetLocalIP gets the non-loopback IP address of this host.
+func GetLocalIP() (string, error) {
+	addrs, err := net.InterfaceAddrs()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	for _, addr := range addrs {
+		var ip net.IP
+		switch v := addr.(type) {
+		case *net.IPNet:
+			ip = v.IP
+		case *net.IPAddr:
+			ip = v.IP
+		default:
+			continue
+		}
+		if !ip.IsLoopback() && ip.IsPrivate() {
+			return ip.String(), nil
+		}
+	}
+	return "", trace.NotFound("No non-loopback local IP address found")
+}
+
+type TestServerOption func(*testing.T, *httptest.Server)
+
+func WithTestServerAddress(ip string) TestServerOption {
+	return func(t *testing.T, srv *httptest.Server) {
+		// Replace the test server's address.
+		l, err := net.Listen("tcp", ip+":0")
```

Comment:
We've had flakiness due to this in the past.

You can't (safely) specify a port of 0 to let the OS pick a port for you, close the listener, and then assume that the port that was previously available will remain available.